lib/phoenix_test/case.ex

```diff
@@ -66,97 +54,65 @@ defmodule PhoenixTest.Case do
 
     alias PhoenixTest.Playwright.Browser
     alias PhoenixTest.Playwright.BrowserContext
+    alias PhoenixTest.Playwright.Page
     alias PhoenixTest.Playwright.Port
 
     @includes_ecto Code.ensure_loaded?(Ecto.Adapters.SQL.Sandbox) &&
                      Code.ensure_loaded?(Phoenix.Ecto.SQL.Sandbox)
 
     def launch_browser(opts) do
       ensure_started()
-      browser = Keyword.fetch!(opts, :browser)
+      {browser, opts} = Keyword.pop!(opts, :browser)
       browser_id = launch_browser(browser, opts)
       on_exit(fn -> post(guid: browser_id, method: :close) end)
       browser_id
     end
 
-    def new_session(%{browser_id: browser_id} = context) do
-      params = if ua = checkout_ecto_repos(context[:async]), do: %{user_agent: ua}, else: %{}
-      context_id = Browser.new_context(browser_id, params)
-      subscribe(context_id)
+    def new_session(context) do
+      browser_context_opts = if ua = checkout_ecto_repos(context[:async]), do: %{user_agent: ua}, else: %{}
+      browser_context_id = Browser.new_context(context.browser_id, browser_context_opts)
+      subscribe(browser_context_id)
 
-      page_id = BrowserContext.new_page(context_id)
-
-      post(%{
-        method: :update_subscription,
-        guid: page_id,
-        params: %{event: :console, enabled: true}
-      })
+      page_id = BrowserContext.new_page(browser_context_id)
+      Page.update_subscription(page_id, event: :console, enabled: true)
 
       frame_id = initializer(page_id).main_frame.guid
-      on_exit(fn -> post(guid: context_id, method: :close) end)
-
-      if trace = context[:trace] do
-        BrowserContext.start_tracing(context_id)
-
-        dir =
-          :phoenix_test
-          |> Application.fetch_env!(:playwright)
-          |> Keyword.get(:trace_dir, "tmp")
-
-        File.mkdir_p!(dir)
-
-        "Elixir." <> module = to_string(context.module)
-        session_id = System.unique_integer([:positive, :monotonic])
+      on_exit(fn -> post(guid: browser_context_id, method: :close) end)
 
-        file =
-          String.replace("#{module}.#{context.test}_#{session_id}.zip", ~r/[^a-zA-Z0-9 \.]/, "_")
+      if context.trace, do: trace(browser_context_id, context)
+      if context.screenshot, do: screenshot(page_id, context)
```

Comment:
Are `:trace` and `:screenshot` guaranteed to exist?

Reply:
Yes, because in `setup_all` we're returning `Config.parse` result (if playwright is enabled) which adds default config values.